### PR TITLE
ENH: Support sklearn metrics in MLHandler  

### PR DESCRIPTION
**Pull request by prakrutisingh24**
MLHandler now supports any scoring metric in Scikit-learn.
Metrics can be specified by users using the `_metric` argument, as follows:
```bash
curl -X POST -F "file=@test.csv" 'http://localhost:9988/?_action=score&_metric=f1'
```
Any metrics listed [here](https://scikit-learn.org/stable/modules/model_evaluation.html) are supported.

### CODE REVIEW

**Comment by jaidevd** [comment]
Thanks a lot @prakrutisingh24, this looks great!

I'd suggest just a few changes on the code to make this more usable.

File: gramex/handlers/mlhandler.py [hunk 1]
```diff
@@ -148,7 +158,7 @@ def set_opt(cls, key, value):
             transform[key] = value
             cls.config_store.dump('transform', transform)
             cls.config_store.update['transform'] = transform
-        elif key in ('class', 'params'):
+        elif key in ('class', 'params', 'score_pref'):
             model = cls.config_store.load('model', {})
```
Comment (jaidevd) [inline gramex/handlers/mlhandler.py:162]: @prakrutisingh24 All of this code in the setup function is running only once at the beginning, i.e. when the model is being _trained_. Scoring the model doesn't make sense here. Scoring actually happens only later when the user sends a POST request. Therefore, all of our scoring code should happen in that function. We can remove all this code.

File: gramex/handlers/mlhandler.py [hunk 2]
```diff
@@ -261,14 +271,56 @@ def _transform(self, data, **kwargs):
...
+                return r2_score(target, predict)
+
+            else:
+                return self.model.score(data,target)
+
+        except ValueError:
+            print('ValueError occured. Please check if the scoring metric is appropriate for the given data!')
+
+
+
```
Comment (jaidevd) [inline gramex/handlers/mlhandler.py:313]: Please remove all the print statements here.

Comment (jaidevd) [reply gramex/handlers/mlhandler.py:313]: Also, the way this function is written, we are restricting the user to choosing between the scoring methods imported at the top of the file.

The intention is to let the user select **any** scoring method, and if it's not applicable, raise an error. We can take any string specified by the user, and make a scorer function out of it: [see this function](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.get_scorer.html). It can be used as follows:

```python
from sklearn.metrics import get_scorer
user_defined_metric = 'accuracy'  # or roc_auc_score, r2, etc.
scorer = get_scorer(user_defined_metric)
accuracy = scorer(model, data, target)
```

File: gramex/handlers/mlhandler.py [hunk 3]
```diff
@@ -261,14 +271,56 @@ def _transform(self, data, **kwargs):
...
             data = self._parse_data(False)
         data = self._transform(data, drop_duplicates=False)
         self.model = cache.open(self.model_path, joblib.load)
         try:
             target = data.pop(score_col)
-            return self.model.score(data, target)
+            s = self._chooseScore(data, target)
+            print('s: ',s)
+            return s
         except KeyError:
```
Comment (jaidevd) [inline gramex/handlers/mlhandler.py:324]: Here, the user will be sending the scoring metric as a request parameter, named `_metric`. So we can use the snippet above to change this.

**Review by jaidevd — CHANGES_REQUESTED** [review]

File: gramex/handlers/mlhandler.py [hunk 4]
```diff
@@ -103,12 +104,13 @@ def setup(cls, data=None, model={}, config_dir='', **kwargs):
 
         cls.set_opt('class', model.get('class'))
         cls.set_opt('params', model.get('params', {}))
+        target_col = cls.get_opt('target_col')
```
Comment (jaidevd) [inline gramex/handlers/mlhandler.py:107]: @prakrutisingh24 This line is not required.

File: gramex/handlers/mlhandler.py [hunk 5]
```diff
@@ -103,12 +104,13 @@ def setup(cls, data=None, model={}, config_dir='', **kwargs):
...
 
         if op.exists(cls.model_path):  # If the pkl exists, load it
             cls.model = joblib.load(cls.model_path)
         elif data is not None:
             mclass = cls.get_opt('class', model.get('class', False))
-            params = cls.get_opt('params', {})
+            params = cls.get_opt('params', {})
```
Comment (jaidevd) [inline gramex/handlers/mlhandler.py:113]: Whitespace at the end of the line, please remove it.

File: gramex/handlers/mlhandler.py [hunk 6]
```diff
@@ -261,14 +263,18 @@ def _transform(self, data, **kwargs):
...
     def _predict(self, data=None, score_col=''):
         if data is None:
             data = self._parse_data(False)
         data = self._transform(data, drop_duplicates=False)
         self.model = cache.open(self.model_path, joblib.load)
         try:
             target = data.pop(score_col)
-            return self.model.score(data, target)
+            metric = self.get_argument('_metric')
+            print('Metric: ',metric)
```
Comment (jaidevd) [inline gramex/handlers/mlhandler.py:275]: Please remove the print statement.

File: gramex/handlers/mlhandler.py [hunk 7]
```diff
@@ -261,14 +263,18 @@ def _transform(self, data, **kwargs):
...
+
     def _predict(self, data=None, score_col=''):
         if data is None:
             data = self._parse_data(False)
         data = self._transform(data, drop_duplicates=False)
         self.model = cache.open(self.model_path, joblib.load)
         try:
             target = data.pop(score_col)
-            return self.model.score(data, target)
+            metric = self.get_argument('_metric')
```
Comment (jaidevd) [inline gramex/handlers/mlhandler.py:274]: If the user does not provide a metric, we need to fall back on the original model of scoring. Please provide for that case also.

**Review by jaidevd — CHANGES_REQUESTED** [review]
@prakrutisingh24 Almost done, just one minor change is left.

File: env/pyvenv.cfg [hunk 8]
```diff
@@ -0,0 +1,3 @@
+home = /Users/prakruti/opt/anaconda3/bin
+include-system-site-packages = false
+version = 3.8.5
```
Comment (jaidevd) [inline env/pyvenv.cfg:3]: Please remove this file.

File: gramex/handlers/mlhandler.py [hunk 9]
```diff
@@ -62,6 +63,7 @@ class MLHandler(FormHandler):
 
     @classmethod
     def setup(cls, data=None, model={}, config_dir='', **kwargs):
+    	print('hellos')
```
Comment (jaidevd) [inline gramex/handlers/mlhandler.py:66]: Please remove this line.

File: gramex/handlers/mlhandler.py (hunk 7 above)
Comment (jaidevd) [reply gramex/handlers/mlhandler.py:273]: This line will cause an error if `_metric` is not provided in the URL. The right way to handle this is as follows:

```python
metric = self.get_argument('_metric', False)  # Note that the second argument is returned if the input is not present
if metric:
    scorer = get_scorer(metric)
    return scorer(self.model, data, target)
return self.model.score(data, target)
```

**Review by jaidevd — CHANGES_REQUESTED** [review]

**Comment by jaidevd** [comment]
@prakrutisingh24 You seem to have accidentally added two unnecessary files. The only file affected by this change should be `gramex/handlers/mlhandler.py`

File: gramex/handlers/mlhandler.py (hunk 5 above)
Comment (jaidevd) [reply gramex/handlers/mlhandler.py:111]: This is still not resolved.

**Review by jaidevd — CHANGES_REQUESTED** [review]

**Comment by jaidevd** [comment]
@sanand0 Can you take a look at this?

**Review by jaidevd — APPROVED** [review]

**Comment by sanand0** [comment]
Thanks @prakrutisingh24
